File: org.palladiosimulator.blockchainsystems.doublespending/src/main/java/org/palladiosimulator/blockchainsystems/doublespending/simulation/termination/BTOState.java
package org.palladiosimulator.blockchainsystems.doublespending.simulation.termination;

import java.util.HashMap;

import org.palladiosimulator.blockchainsystems.core.system.abstractions.Block;
import org.palladiosimulator.blockchainsystems.core.system.abstractions.BlockType;

public class BTOState {

    private final BlockState _btoState;
    private final HashMap<String, BlockState> _btouSuccessorsByHashes;

    private final int _numberOfBlocksRequiredForBTOAcceptance;

    private boolean _isBTOAccepted;
    private boolean _wasBTOAccepted;

    public BTOState(BlockState btoState, int numberOfBlocksRequiredForBTOAcceptance) {
        _btoState = btoState;
        _numberOfBlocksRequiredForBTOAcceptance = numberOfBlocksRequiredForBTOAcceptance;

        _btouSuccessorsByHashes = new HashMap<String, BlockState>();

        _isBTOAccepted = false;
        _wasBTOAccepted = false;
    }

    public boolean isBTOAccepted() {
        return _isBTOAccepted;
    }

    public boolean wasBTOAccepted() {
        return _wasBTOAccepted;
    }

    public boolean isBlockSuccessorOfBTO(Block block) {
        return _btoState.getBlock().getHash() == block.getPreviousHash()
                || _btouSuccessorsByHashes.containsKey(block.getPreviousHash());
    }

    public void addBTOSuccessorBlock(BlockState blockState) {
        if (!isBlockSuccessorOfBTO(blockState.getBlock())) {
            return;
        }

        _btouSuccessorsByHashes.put(blockState.getBlock().getHash(), blockState);

        updateIsBTOAccepted();
    }

    public void onBTOBlockTypeChanged(BlockType newBlockType) {
        _btoState.setBlockType(newBlockType);

        updateIsBTOAccepted();
    }

    private void updateIsBTOAccepted() {
        if (isBTOIncluded() && hasRequiredLongestChainSuccessorForAcceptance()) {
            _isBTOAccepted = true;
            if (!_wasBTOAccepted) {
                _wasBTOAccepted = true;
            }
        } else {
            _isBTOAccepted = false;
        }
    }

    private boolean hasRequiredLongestChainSuccessorForAcceptance() {
        return getNumberOfLongestChainSuccessors() >= _numberOfBlocksRequiredForBTOAcceptance;
    }

    public boolean isBTOIncluded() {
        return _btoState.getBlockType() == BlockType.IncludedBlock;
    }

    public long getNumberOfLongestChainSuccessors() {
        if (_btouSuccessorsByHashes.isEmpty()) {
            return 0;
        }

        long latestSuccessorPosition = _btouSuccessorsByHashes
                .values()
                .stream()
                .mapToLong(BlockState::getBlockPosition)
                .max()
                .getAsLong();

        return latestSuccessorPosition - _btoState.getBlockPosition();
    }

    public BlockState getBTO() {
        return _btoState;
    }
}

File: org.palladiosimulator.blockchainsystems.trilemma/src/main/kotlin/org/palladiosimulator/blockchainsystems/trilemma/simulation/termination/BTOState.kt
package org.palladiosimulator.blockchainsystems.trilemma.simulation.termination

import org.palladiosimulator.blockchainsystems.core.system.abstractions.Block
import org.palladiosimulator.blockchainsystems.core.system.abstractions.BlockType

class BTOState(
  private val btoState: BlockState,
  private val numberOfBlocksRequiredForBTOAcceptance: Int
) {
  private val btouSuccessorsByHashes = HashMap<String, BlockState>()

  var isBTOAccepted: Boolean = false
    private set

  var wasBTOAccepted: Boolean = false
    private set

  fun isBlockSuccessorOfBTO(block: Block): Boolean {
    return btoState.block.getHash() === block.getPreviousHash()
      || btouSuccessorsByHashes.containsKey(block.getPreviousHash())
  }

  fun addBTOSuccessorBlock(blockState: BlockState) {
    if (!isBlockSuccessorOfBTO(blockState.block)) return
    btouSuccessorsByHashes.put(blockState.block.getHash(), blockState)
    updateIsBTOAccepted()
  }

  fun onBTOBlockTypeChanged(newBlockType: BlockType) {
    btoState.blockType = newBlockType
    updateIsBTOAccepted()
  }

  private fun updateIsBTOAccepted() {
    if (isBTOIncluded && hasRequiredLongestChainSuccessorForAcceptance()) {
      isBTOAccepted = true
      if (!wasBTOAccepted) wasBTOAccepted = true
    } else {
      isBTOAccepted = false
    }
  }

  private fun hasRequiredLongestChainSuccessorForAcceptance(): Boolean {
    return this.numberOfLongestChainSuccessors >= numberOfBlocksRequiredForBTOAcceptance
  }

  val isBTOIncluded: Boolean
    private set
    get() = btoState.blockType == BlockType.IncludedBlock

  val numberOfLongestChainSuccessors: Long
    private set
    get() {
      if (btouSuccessorsByHashes.isEmpty()) return 0

      val latestSuccessorPosition = btouSuccessorsByHashes
        .values
        .stream()
        .mapToLong { obj: BlockState -> obj.blockPosition }
        .max()
        .getAsLong()

      return latestSuccessorPosition - btoState.blockPosition
    }

  val bto: BlockState
    private set
    get() = btoState
}